LamestWebserver/LamestWebserver/Core/Logger.cs

```diff
@@ -259,28 +336,55 @@ public void CrashAndBurn(string msg)
             Environment.Exit(-1);
         }
 
-        internal void Log(ELoggingLevel loggingLevel, string msg)
+        private void Log(ELoggingLevel loggingLevel, string msg)
         {
-            // TODO: Implement propper logging!
-
-            ServerHandler.LogMessage($"[{loggingLevel.ToString()} \\\\\\\\ {msg}]");
+            if (_currentOutputSource != EOutputSource.None)
+            {
+                using (_loggerMutex.Lock())
+                {
+                    _multiStreamWriter.WriteLine($"[{loggingLevel.ToString()} \\\\\\\\ {msg}]");
```

Comment:
what happens, if a stream hasn't been initialized or is disposed? (not saying you haven't already covered that - but I'm just wondering)

Reply:
there is a System.ObjectDisposedException if you write in a closed file but this only happens if you manual close the stream without opening it again

Reply:
What happens if you write, when another thread runs over [this](https://github.com/rainerzufalldererste/LamestWebserver/pull/18/files/18bea505ee62066981de40cba3b8d558a6b6e006#diff-2969595700fb2803c8f89354eff12b12R62) line at the same time?

Reply:
so that the stream was just closed

Reply:
I could make one Atomic function out of this so closing and opening is locked by one mutex

Reply:
I'm not sure on this, but: some synchronizers like `ReadWriteLock` have recursive locking, so you could just use one of them and lock+release twice in the specified method that I linked to.

``` c#
void x()
{
using(_writeLock.LockWrite())
{
a(); // locks _writeLock internally for writing.
b(); // locks _writeLock internally for writing.
}
}
```

this may also be possible with *some* mutexes afaik...

Reply:
👍